source/server/hot_restarting_parent.cc

```diff
@@ -0,0 +1,127 @@
+#include "server/hot_restarting_parent.h"
+
+#include "envoy/server/instance.h"
+
+#include "common/network/utility.h"
+
+namespace Envoy {
+namespace Server {
+
+HotRestartingParent::HotRestartingParent(int base_id, int restart_epoch)
+    : HotRestartingBase(base_id), restart_epoch_(restart_epoch) {
+  child_address_ = createDomainSocketAddress(restart_epoch_ + 1, "child");
+  bindDomainSocket(restart_epoch_, "parent");
+}
+
+void HotRestartingParent::initialize(Event::Dispatcher& dispatcher, Server::Instance& server) {
+  socket_event_ =
+      dispatcher.createFileEvent(my_domain_socket(),
+                                 [this](uint32_t events) -> void {
+                                   ASSERT(events == Event::FileReadyType::Read);
+                                   onSocketEvent();
+                                 },
+                                 Event::FileTriggerType::Edge, Event::FileReadyType::Read);
+  server_ = &server;
+}
+
+void HotRestartingParent::onGetListenSocket(RpcGetListenSocketRequest& rpc) {
+  RpcGetListenSocketReply reply;
+  reply.fd_ = -1;
+
+  Network::Address::InstanceConstSharedPtr addr =
+      Network::Utility::resolveUrl(std::string(rpc.address_));
+  for (const auto& listener : server_->listenerManager().listeners()) {
+    if (*listener.get().socket().localAddress() == *addr) {
+      reply.fd_ = listener.get().socket().ioHandle().fd();
+      break;
+    }
+  }
+
+  if (reply.fd_ == -1) {
+    // In this case there is no fd to duplicate so we just send a normal message.
+    sendMessage(child_address_, reply);
```

Comment:
The one functional change. Since each HotRestarting{Child,Parent} has one of these, and we have one of each, we now have two datagram domain sockets: one for talking to the parent, one for eventually talking to the child. Their sockaddr_un addresses are now differentiated: they include the word "child" if the socket with that address is being used by the child half of the logic, "parent" otherwise.